tests/test_cli_compile.py

```diff
@@ -4044,3 +4044,63 @@ def test_second_order_requirements_relative_path_in_separate_dir(
             # via -r {output_path}
         """
     )
+
+
+@pytest.mark.parametrize("pyproject_path_is_absolute", (True, False))
+def test_that_self_referential_pyproject_toml_extra_can_be_compiled(
+    pip_conf, runner, tmp_path, monkeypatch, pyproject_path_is_absolute
+):
+    """
+    Test that a `pyproject.toml` source file can use self-referential extras
+    which point back to the original package name.
+
+    This is a regression test for:
+    https://github.com/jazzband/pip-tools/issues/2215
+    """
+    src_file = tmp_path / "pyproject.toml"
+    src_file.write_text(
+        dedent(
+            """
+            [project]
+            name = "foo"
+            version = "0.1.0"
+            [project.optional-dependencies]
+            ext1 = ["small-fake-a"]
+            ext2 = ["foo[ext1]"]
+            """
+        )
+    )
+
+    if pyproject_path_is_absolute:
+        input_path = src_file.as_posix()
+    else:
+        input_path = src_file.relative_to(tmp_path).as_posix()
+
+    with monkeypatch.context() as revertable_ctx:
+        revertable_ctx.chdir(tmp_path)
+        out = runner.invoke(
+            cli,
+            [
+                "--output-file",
+                "-",
+                "--quiet",
+                "--no-header",
+                "--no-emit-options",
+                # use in-process setuptools
+                "--no-build-isolation",
+                # importantly, request the extra which uses the self-reference
+                "--extra",
+                "ext2",
+                input_path,
+            ],
+        )
+
+    assert out.exit_code == 0
+    assert out.stdout == dedent(
+        f"""\
+        foo[ext1] @ {src_file.parent.absolute().as_uri()}
```

Comment:
Is this really what we want? I don't think people would expect absolute paths from someone else's computer to end up in their lock files..

Reply:
I agree, and it's why I was pretty hesitant at first. But also, that's the behavior from 7.4.1 -- it's just not tested.

In theory, someone could be using this in some controlled context like an image build and _wanting_ the concrete dependency path that this provides. So although I'm uncertain that it's right for the long term, I don't want to treat it as blocking the fix.

Reply:
Wait, did it always use absolute paths in 7.4.1? Not relative? I thought it'd conform to the shape of input that the 7.5.0 changelog promises.. Can we use `file://./rel/path/`?

Reply:
I suppose, we could postpone that for now. There might be a use case coverage missing for stripping the `@ http://..source` in some contexts.